src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -350,39 +351,104 @@ private void WriteAsPaxSharedInternal(Span<byte> buffer, out long actualLength)
+            int encoded = Encoding.UTF8.GetBytes(truncatedName, destination);
+
+            return Checksum(destination.Slice(0, encoded));
         }
 
-        // Ustar and PAX save in the name byte array only the ASCII bytes that fit, and the rest of that string is saved in the prefix field.
-        private int WritePosixName(Span<byte> buffer)
+        // 'https://www.freebsd.org/cgi/man.cgi?tar(5)'
+        // If the pathname is too long to fit in the 100 bytes provided by the standard format,
+        // it can be split at any / character with the first portion going into the prefix field.
+        private int WriteUstarName(Span<byte> buffer)
         {
-            int checksum = WriteName(buffer);
+            const int MaxPathname = FieldLengths.Prefix + 1 + FieldLengths.Name;
+            if (Encoding.UTF8.GetByteCount(_name) > MaxPathname)
+            {
+                throw new ArgumentException(SR.TarEntryNameExceedsMaxLength, ArgNameEntry);
+            }
+
+            Span<byte> encodingBuffer = stackalloc byte[MaxPathname];
+            int encoded = Encoding.UTF8.GetBytes(_name, encodingBuffer);
+            ReadOnlySpan<byte> nameAndPrefixBytes = encodingBuffer.Slice(0, encoded);
 
-            if (_name.Length > FieldLengths.Name)
+            if (nameAndPrefixBytes.Length <= FieldLengths.Name)
             {
-                int prefixBytesLength = Math.Min(_name.Length - FieldLengths.Name, FieldLengths.Prefix);
-                Span<byte> remaining = stackalloc byte[prefixBytesLength];
-                int encoded = Encoding.ASCII.GetBytes(_name.AsSpan(FieldLengths.Name, prefixBytesLength), remaining);
-                Debug.Assert(encoded == remaining.Length);
+                return WriteLeftAlignedBytesAndGetChecksum(nameAndPrefixBytes, buffer.Slice(FieldLocations.Name, FieldLengths.Name));
+            }
 
-                checksum += WriteLeftAlignedBytesAndGetChecksum(remaining, buffer.Slice(FieldLocations.Prefix, FieldLengths.Prefix));
+            int lastIdx = nameAndPrefixBytes.LastIndexOfAny(SeparatorsAsBytes);
+            ReadOnlySpan<byte> name = stackalloc byte[0];
+            ReadOnlySpan<byte> prefix = stackalloc byte[0];
+
+            if (lastIdx == -1)
+            {
+                name = nameAndPrefixBytes;
+                prefix = default;
             }
+            else
+            {
+                name = nameAndPrefixBytes.Slice(lastIdx + 1);
 
-            return checksum;
+                if (lastIdx > 0)
+                {
+                    prefix = nameAndPrefixBytes.Slice(0, lastIdx);
+                }
+                else
+                {
+                    // We cannot neglect the separator if that's the only thing we can place in prefix.
+                    prefix = nameAndPrefixBytes.Slice(0, 1);
+                }
+            }
+
+            while (prefix.Length - name.Length > FieldLengths.Prefix)
```

Comment:
Can you add a comment as to what this loop is doing?  It seems like it's trying to get the prefix to be short enough to fit in the prefix field?

Reply:
yes, that's what it attempts and it gives up if the prefix would be > 155.